engine/consensus/approvals/core.go

```diff
@@ -0,0 +1,336 @@
+	state                                protocol.State                  // used to access protocol state
+	verifier                             module.Verifier                 // used to validate result approvals
+	seals                                mempool.IncorporatedResultSeals // holds candidate seals for incorporated results that have acquired sufficient approvals; candidate seals are constructed  without consideration of the sealability of parent results
+	approvalConduit                      network.Conduit                 // used to request missing approvals from verification nodes
+	requestTracker                       *sealing.RequestTracker         // used to keep track of number of approval requests, and blackout periods, by chunk
+}
+
+func NewApprovalProcessingCore(headers storage.Headers, state protocol.State, assigner module.ChunkAssigner,
+	verifier module.Verifier, seals mempool.IncorporatedResultSeals, approvalConduit network.Conduit, requiredApprovalsForSealConstruction uint, emergencySealingActive bool) (*approvalProcessingCore, error) {
+
+	lastSealed, err := state.Sealed().Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve last sealed block: %w", err)
+	}
+
+	core := &approvalProcessingCore{
+		approvalsCache:                       NewApprovalsCache(1000),
+		assigner:                             assigner,
+		headers:                              headers,
+		state:                                state,
+		verifier:                             verifier,
+		seals:                                seals,
+		approvalConduit:                      approvalConduit,
+		requiredApprovalsForSealConstruction: requiredApprovalsForSealConstruction,
+		emergencySealingActive:               emergencySealingActive,
+		requestTracker:                       sealing.NewRequestTracker(10, 30),
+	}
+
+	factoryMethod := func(result *flow.ExecutionResult) (*AssignmentCollector, error) {
+		return NewAssignmentCollector(result, core.state, core.headers, core.assigner, core.seals, core.verifier,
+			core.approvalConduit, core.requestTracker, core.requiredApprovalsForSealConstruction)
+	}
+
+	collectors := NewAssignmentCollectorTree(factoryMethod)
+	_, err = collectors.PruneUpToHeight(lastSealed.Height)
+	if err != nil {
+		return nil, fmt.Errorf("could not prune tree to initial height")
+	}
+
+	core.collectorTree = collectors
+
+	return core, nil
+}
+
+func (c *approvalProcessingCore) lastSealedHeight() uint64 {
+	return atomic.LoadUint64(&c.atomicLastSealedHeight)
+}
+
+func (c *approvalProcessingCore) lastFinalizedHeight() uint64 {
+	return atomic.LoadUint64(&c.atomicLastFinalizedHeight)
+}
+
+// WARNING: this function is implemented in a way that we expect blocks strictly in parent-child order
+// Caller has to ensure that it doesn't feed blocks that were already processed or in wrong order.
+func (c *approvalProcessingCore) OnFinalizedBlock(finalizedBlockID flow.Identifier) {
```

Comment:
This method should not block (otherwise, it will block HotStuff). But without blocking HotStuff, we loose our order guarantee. Hence, I would prefer if this method could process finalization out of order. Specifically:
* finalization requests for already finalized heights can be skipped
* implementation logic should support skipping values